framework/src/multiapps/MultiApp.C

```diff
@@ -421,10 +429,58 @@ MultiApp::restore()
   if (_apps.empty())
     return;
 
+  // We temporatily copy and store solutions for all subapps
+  if (_keep_solution_during_restore)
+  {
+    _end_solutions.resize(_my_num_apps);
+
+    for (unsigned int i = 0; i < _my_num_apps; i++)
+    {
+      _end_solutions[i] =
+          _apps[i]->getExecutioner()->feProblem().getNonlinearSystemBase().solution().clone();
+      auto & sub_multiapps =
+          _apps[i]->getExecutioner()->feProblem().getMultiAppWarehouse().getObjects();
+
+      // Override subapp only when _keep_solution_during_restore is true
+      if (_keep_solution_during_restore)
+      {
+        // multiapps of each subapp should do the same things
+        // It is implemented recursively
+        for (auto & multi_app : sub_multiapps)
+          multi_app->keepSolutionDuringRestore(_keep_solution_during_restore);
+      }
```

Comment:
Should this be outside of the if statement in 433? If not, then you can remove the if statement here.